tests/fpgaperf/test_fpgaperf_counter_c.cpp

```diff
@@ -0,0 +1,192 @@
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif // HAVE_CONFIG_H
+
+#include "fpgaperf_counter.h"
+
+extern "C" {
+
+#include <fcntl.h>
+#include <glob.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <libudev.h>
+#include <linux/perf_event.h>
+
+#define DFL_PERF_STR_MAX        256
+
+
+/* parse the events and format value using sysfs_path */
+fpga_result fpga_perf_events(char* perf_sysfs_path, fpga_perf_counter *fpga_perf);
+}
+
+#include "intel-fpga.h"
+#include <linux/ioctl.h>
+
+#include <cstdlib>
+#include <string>
+
+#include <opae/fpga.h>
+#include <opae/properties.h>
+#include <opae/utils.h>
+
+#include "gtest/gtest.h"
+#include "mock/test_system.h"
+
+
+using namespace opae::testing;
+
+class fpgaperf_counter_c_p : public ::testing::TestWithParam<std::string> {
+protected:
+        fpgaperf_counter_c_p(): tokens_{ {nullptr, nullptr} } {}
+
+        virtual void SetUp() override 
+        {
+		ASSERT_TRUE(test_platform::exists(GetParam()));
+		platform_ = test_platform::get(GetParam());
+		system_ = test_system::instance();
+		system_->initialize();
+		system_->prepare_syfs(platform_);
+
+		filter_ = nullptr;
+		ASSERT_EQ(fpgaInitialize(NULL), FPGA_OK);
+		ASSERT_EQ(fpgaGetProperties(nullptr, &filter_), FPGA_OK);
+		ASSERT_EQ(fpgaPropertiesSetObjectType(filter_, FPGA_DEVICE), FPGA_OK);
+		num_matches_ = 0;
+		ASSERT_EQ(fpgaEnumerate(&filter_, 1, tokens_.data(), tokens_.size(),
+			&num_matches_), FPGA_OK);
+		EXPECT_GT(num_matches_, 0);
+		dev_ = nullptr;
+		ASSERT_EQ(fpgaOpen(tokens_[0], &dev_, 0), FPGA_OK);
+		fpga_perf = new fpga_perf_counter;
+		EXPECT_EQ(fpgaPerfCreateHandle(fpga_perf), FPGA_OK);
+        }
+
+        virtual void TearDown() override {
+		EXPECT_EQ(fpgaPerfDestroyHandle(fpga_perf), FPGA_OK);
+		EXPECT_EQ(fpgaDestroyProperties(&filter_), FPGA_OK);
+		if (dev_) {
+			EXPECT_EQ(fpgaClose(dev_), FPGA_OK);
+			dev_ = nullptr;
+		}
+		for (auto &t : tokens_) {
+			if (t) {
+				EXPECT_EQ(fpgaDestroyToken(&t), FPGA_OK);
+				t = nullptr;
+			}
+		}
+		delete fpga_perf;
```

Comment:
If the assertion on line 88 fails then line 89 won't execute (so we shouldn't delete)